libs/nest/config/src/lib/ConfigurationLoader.ts

```diff
@@ -0,0 +1,162 @@
+import { ConfigDefinition, Configuration, EnvLoader } from './types'
+import { ServerSideFeatureClient } from '@island.is/feature-flags'
+import { logger } from '@island.is/logging'
+import { InvalidConfiguration } from './InvalidConfiguration'
+import { ConfigurationError } from './ConfigurationError'
+import { ZodError } from 'zod'
+
+type Issue = 'missing' | 'could not be parsed as JSON'
+
+export class ConfigurationLoader<T> implements EnvLoader {
+  private isDev = process.env.NODE_ENV !== 'production'
+  private issues: Record<string, Issue> = {}
+
+  constructor(private definition: ConfigDefinition<T>) {}
+
+  load(): Configuration<T> {
+    if (
+      this.definition.serverSideFeature &&
+      !ServerSideFeatureClient.isOn(this.definition.serverSideFeature)
+    ) {
+      logger.info({
+        category: 'ConfigModule',
+        message: `Ignored configuration for ${this.definition.name}. Server-side feature flag missing: ${this.definition.serverSideFeature}`,
+      })
+      return new InvalidConfiguration(
+        `Unable to read configuration for ${this.definition.name}. Server-side feature flag missing: ${this.definition.serverSideFeature}`,
+      ) as Configuration<T>
+    }
+
+    try {
+      const result = this.definition.load(this)
+
+      // Loader did not crash, but there might still be issues we need to deal with.
+      return this.handleResult(result)
+    } catch (err) {
+      // Loader crashed. If there are reported issues, throw those instead.
+      return this.handleResult(undefined, err)
+    }
+  }
+
+  private handleResult(result?: T, error?: Error) {
+    const issues = Object.entries(this.issues)
+    const errorMessage = this.formatIssues(issues)
+    const hasParseIssue = issues.find(
+      (issue) => issue[1] === 'could not be parsed as JSON',
+    )
+    const hasMissingIssue = !hasParseIssue && issues.length > 0
+
+    // Always throw if there are any parse issues.
+    if (hasParseIssue) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Only throw missing issue in production
+    if (hasMissingIssue && !this.isDev) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Re-throw other unexpected errors from loader.
+    if (error) {
+      throw error
+    }
+
+    if (result === undefined) {
+      throw new ConfigurationError(
+        `Failed loading configuration for ${this.definition.name}. Config was undefined.`,
+      )
+    }
+
+    if (this.isDev && hasMissingIssue) {
+      logger.warn({
+        message: `Could not load configuration for ${this.definition.name}. Missing ${issues.length} required environment variable(s).`,
+        category: 'ConfigModule',
+      })
+      return new InvalidConfiguration(errorMessage) as Configuration<T>
```

Comment:
It seems like it would be simpler to throw, right away, but I have a feeling there is a particular use case you were addressing by throwing on property access, rather than at config time, I'd be curious to know what it is

Reply:
Right! It's to support partial configuration during local development, especially for the `api` and `application-system-api` projects.

When you're working on something, you should only need the dev-secrets (environment variables) and services for that thing on your computer. Otherwise, developers would need to have all of the secrets, for all of the modules, just to start these services (vmst, driver-license, health-insurance, etc).

I didn't want all of the required configuration attributes to be nullable to deal with this issue locally, as well as missing server-side features in production. Instead you can check `isConfigured`, or deal with the error at runtime. If you end up accessing an API that reads a missing configuration, you'll get a verbose error telling you precisely what 
the issue is (eg which environment variable which is missing or which server side feature is disabled). 

The only problem with this implementation is that in many cases, the configuration is read at server startup (eg when passed to OpenAPI classes or `createEnhancedFetch`). It could be tricky to delay this initialization until a request needs it.